diffusion/algorithms/discriminator_schedule.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2022 MosaicML Diffusion authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Scheduler algorithm for a discriminator."""
+import torch.nn as nn
+from composer.core import Algorithm, Event, State, Time, TimeUnit
+from composer.loggers import Logger
+
+__all__ = ['DiscriminatorSchedule']
+
+
+class DiscriminatorSchedule(Algorithm):
+    """Scheduling algorithm for the autoencoder discriminator.
+
+    Args:
+        start_iteration (str): Time string specifying the iteration (in units of batches) to start training the
+            discriminator. Default: `'0ba'`.
+    """
+
+    def __init__(self, start_iteration: str = '0ba') -> None:
+        self.start_iteration = Time.from_timestring(start_iteration)
+        self.lr = 0.0
+        self.weight_decay = 0.0
+        self.discriminator_weight = 0.0
+
+        # Verify the start iteration is in units of batches
+        if self.start_iteration.unit != TimeUnit.BATCH:
+            raise ValueError('start_iteration must be in units of batches')
+
+    def match(self, event: Event, state: State) -> bool:
+        return event in [Event.INIT, Event.BATCH_START, Event.AFTER_LOSS]
+
+    def apply(self, event: Event, state: State, logger: Logger) -> None:
+        if hasattr(state.model, 'autoencoder_loss') and isinstance(state.model.autoencoder_loss, nn.Module):
+            autoencoder_loss = state.model.autoencoder_loss
+        elif isinstance(state.model.module, nn.Module) and hasattr(state.model.module, 'autoencoder_loss'):
```

Comment:
You're correct, there would be an error if the model 1) does not contain an autoencoder loss, and 2) is not DDP wrapped. Case 1 would always cause an error, but case 1 with case 2 would cause the error to not be caught. I fixed it, which required refactoring these if-statements a bit.

Reply:
The changes look great!